File: examples/python/beatapi.py
"""Small dependency-free BeatAPI reference client.

This module is example code, not an officially versioned SDK.
"""

from __future__ import annotations

import json
import os
import random as random_module
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Callable


TERMINAL_STATUSES = {"succeeded", "failed"}


class BeatAPIError(Exception):
    def __init__(
        self,
        message: str,
        *,
        status: int | None = None,
        code: str | None = None,
        request_id: str | None = None,
        details: Any = None,
    ) -> None:
        super().__init__(message)
        self.status = status
        self.code = code
        self.request_id = request_id
        self.details = details


class BeatAPIClient:
    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        transport: Callable[[urllib.request.Request], Any] = urllib.request.urlopen,
        sleep: Callable[[float], None] = time.sleep,
        random: Callable[[], float] = random_module.random,
    ) -> None:
        self.api_key = api_key if api_key is not None else os.getenv("BEATAPI_API_KEY")
        if not self.api_key:
            raise ValueError(
                "BEATAPI_API_KEY is required. Create a key in the BeatAPI dashboard."
            )

        self.base_url = (
            base_url or os.getenv("BEATAPI_BASE_URL") or "https://api.beatapi.io"
        ).rstrip("/")
        self.transport = transport
        self.sleep = sleep
        self.random = random

    def request(
        self,
        path: str,
        *,
        method: str = "GET",
        body: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> Any:
        data = json.dumps(body).encode() if body is not None else None
        request_headers = {
            "Accept": "application/json",
            "Authorization": f"Bearer {self.api_key}",
            **(headers or {}),
        }
        if data is not None:
            request_headers["Content-Type"] = "application/json"

        request = urllib.request.Request(
            f"{self.base_url}{path}",
            data=data,
            headers=request_headers,
            method=method,
        )

        try:
            response = self.transport(request)
            with response:
                status = response.status
                payload = json.loads(response.read().decode() or "{}")
        except urllib.error.HTTPError as error:
            status = error.code
            payload = json.loads(error.read().decode() or "{}")

        if status < 200 or status >= 300:
            public_error = payload.get("error", {})
            raise BeatAPIError(
                public_error.get(
                    "message", f"BeatAPI request failed with HTTP {status}."
                ),
                status=status,
                code=public_error.get("code"),
                request_id=public_error.get("request_id"),
                details=public_error.get("details"),
            )

        return payload.get("data")

    def create_music_video_task(self, input_data: dict[str, Any]) -> dict[str, Any]:
        return self.request(
            "/v1/music-video/tasks",
            method="POST",
            body=input_data,
        )

    def create_ecommerce_video_task(
        self, input_data: dict[str, Any]
    ) -> dict[str, Any]:
        return self.request(
            "/v1/ecommerce-video/tasks",
            method="POST",
            body=input_data,
        )

    def create_realtime_session(
        self,
        input_data: dict[str, Any],
        *,
        idempotency_key: str,
    ) -> dict[str, Any]:
        if not idempotency_key:
            raise ValueError("idempotency_key is required")
        return self.request(
            "/v1/realtime/sessions",
            method="POST",
            body=input_data,
            headers={"Idempotency-Key": idempotency_key},
        )

    def get_realtime_session(self, session_id: str) -> dict[str, Any]:
        encoded = urllib.parse.quote(session_id, safe="")
        return self.request(f"/v1/realtime/sessions/{encoded}")

    def close_realtime_session(self, session_id: str) -> dict[str, Any]:
        encoded = urllib.parse.quote(session_id, safe="")
        return self.request(
            f"/v1/realtime/sessions/{encoded}",
            method="DELETE",
        )

    def get_task(self, task_id: str) -> dict[str, Any]:
        return self.request(f"/v1/tasks/{urllib.parse.quote(task_id, safe='')}")

    def get_usage(self) -> dict[str, Any]:
        return self.request("/v1/usage")

    def wait_for_task(
        self,
        task_id: str,
        *,
        interval_seconds: float = 5,
        max_attempts: int = 120,
        on_update: Callable[[dict[str, Any], int], None] | None = None,
    ) -> dict[str, Any]:
        for attempt in range(1, max_attempts + 1):
            task = self.get_task(task_id)
            if on_update:
                on_update(task, attempt)

            if task["status"] in TERMINAL_STATUSES:
                return task

            if attempt < max_attempts:
                jitter = interval_seconds * 0.2 * self.random()
                self.sleep(interval_seconds + jitter)

        raise TimeoutError(
            f"Task {task_id} did not finish after {max_attempts} attempts."
        )

File: examples/node/lib/beatapi.mjs
const TERMINAL_STATUSES = new Set(["succeeded", "failed"]);

export class BeatAPIError extends Error {
  constructor(message, { status, code, requestId, details } = {}) {
    super(message);
    this.name = "BeatAPIError";
    this.status = status;
    this.code = code;
    this.requestId = requestId;
    this.details = details;
  }
}

export class BeatAPIClient {
  constructor({
    apiKey = process.env.BEATAPI_API_KEY,
    baseUrl = process.env.BEATAPI_BASE_URL || "https://api.beatapi.io",
    fetchImpl = globalThis.fetch,
    sleep = (milliseconds) =>
      new Promise((resolve) => setTimeout(resolve, milliseconds)),
    random = Math.random,
  } = {}) {
    if (!apiKey) {
      throw new Error(
        "BEATAPI_API_KEY is required. Create a key in the BeatAPI dashboard.",
      );
    }
    if (typeof fetchImpl !== "function") {
      throw new Error("A fetch implementation is required.");
    }

    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.fetch = fetchImpl;
    this.sleep = sleep;
    this.random = random;
  }

  async request(path, { method = "GET", body, headers = {} } = {}) {
    const requestHeaders = {
      accept: "application/json",
      authorization: `Bearer ${this.apiKey}`,
      ...headers,
    };
    let requestBody = body;

    if (body !== undefined && !(body instanceof FormData)) {
      requestHeaders["content-type"] = "application/json";
      requestBody = JSON.stringify(body);
    }

    const response = await this.fetch(`${this.baseUrl}${path}`, {
      method,
      headers: requestHeaders,
      body: requestBody,
    });
    const payload = await response.json().catch(() => ({}));

    if (!response.ok) {
      const error = payload.error || {};
      throw new BeatAPIError(
        error.message || `BeatAPI request failed with HTTP ${response.status}.`,
        {
          status: response.status,
          code: error.code,
          requestId: error.request_id,
          details: error.details,
        },
      );
    }

    return payload.data;
  }

  createMusicVideoTask(input) {
    return this.request("/v1/music-video/tasks", {
      method: "POST",
      body: input,
    });
  }

  createEcommerceVideoTask(input) {
    return this.request("/v1/ecommerce-video/tasks", {
      method: "POST",
      body: input,
    });
  }

  createRealtimeSession(input, { idempotencyKey } = {}) {
    if (!idempotencyKey) {
      throw new TypeError("idempotencyKey is required.");
    }
    return this.request("/v1/realtime/sessions", {
      method: "POST",
      body: input,
      headers: { "idempotency-key": idempotencyKey },
    });
  }

  getRealtimeSession(sessionId) {
    return this.request(
      `/v1/realtime/sessions/${encodeURIComponent(sessionId)}`,
    );
  }

  closeRealtimeSession(sessionId) {
    return this.request(
      `/v1/realtime/sessions/${encodeURIComponent(sessionId)}`,
      { method: "DELETE" },
    );
  }

  getTask(taskId) {
    return this.request(`/v1/tasks/${encodeURIComponent(taskId)}`);
  }

  getUsage() {
    return this.request("/v1/usage");
  }

  uploadFile(file, filename = file?.name || "upload.bin") {
    const form = new FormData();
    form.append("file", file, filename);
    return this.request("/v1/files", {
      method: "POST",
      body: form,
    });
  }

  async waitForTask(
    taskId,
    { intervalMs = 5_000, maxAttempts = 120, onUpdate } = {},
  ) {
    for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
      const task = await this.getTask(taskId);
      onUpdate?.(task, attempt);

      if (TERMINAL_STATUSES.has(task.status)) {
        return task;
      }

      if (attempt < maxAttempts) {
        const jitter = Math.floor(intervalMs * 0.2 * this.random());
        await this.sleep(intervalMs + jitter);
      }
    }

    throw new Error(
      `Task ${taskId} did not finish after ${maxAttempts} attempts.`,
    );
  }
}
